src/main/java/org/mockito/ArgumentMatcher.java

```diff
@@ -125,4 +125,44 @@
      * @return true if this matcher accepts the given argument.
      */
     boolean matches(T argument);
+
+    /**
+     * The type of the argument this matcher matches.
+     *
+     * <p>This method is used to differentiate between a matcher used to match a raw vararg array parameter
+     * from a matcher used to match a single value passed as a vararg parameter.
+     *
+     * <p>Where the matcher:
+     * <ul>
+     *     <li>is at the parameter index of a vararg parameter</li>
+     *     <li>is the last matcher passed</li>
+     *     <li>this method returns a type assignable to the vararg parameter's raw type, i.e. its array type.</li>
+     * </ul>
+     *
+     * ...then the matcher is matched against the raw vararg parameter, rather than the first element of the raw parameter.
+     *
+     * <p>For example:
+     *
+     * <pre class="code"><code class="java">
+     *  // Given vararg method with signature:
+     *  int someVarargMethod(String... args);
+     *
+     *  // The following will match invocations with any number of parameters, i.e. any number of elements in the raw array.
+     *  mock.someVarargMethod(isA(String[].class));
+     *
+     *  // The following will match invocations with a single parameter, i.e. one string in the raw array.
+     *  mock.someVarargMethod(isA(String.class));
+     *
+     *  // The following will match invocations with two parameters, i.e. two strings in the raw array
+     *  mock.someVarargMethod(isA(String.class), isA(String.class));
+     * </code></pre>
+     *
+     * <p>Only matcher implementations that can conceptually match a raw vararg parameter should override this method.
+     *
+     * @return the type this matcher handles. The default value of {@link Void} means the type is not known.
+     * @since 4.10.0
+     */
+    default Class<?> type() {
```

Comment:
With just 1 method, `ArgumentMatcher` automatically qualifies as a `FunctionalInterface`. E.g. it can be written as a clean lambda. With the addition of this method, can users still write an `ArgumentMatcher` as a lambda? In case we remove the default implementation, is that also true?

If that is not possible, we can consider splitting up the argument matchers into two. We can keep `ArgumentMatcher` as is, but we define a new interface `ArgumentMatcherWithType` that extends `ArgumentMatcher`, but does not define a default implementation. Then we can implement this interface in all our internal matchers, but external users can keep on using `ArgumentMatcher` as-is.

In the logic below, we would then still check if the `ArgumentMatcher` we have is also `ArgumentMatcherWithType` and apply the logic there. WDYT?

Reply:
Now that I realize it, this comment would basically boil down to: add a new method to `VarargMatcher` that defines its type. That sounds quite natural to me and would be a minimally impactful change, yet still provide significant value by solving all the issues that you linked.

Reply:
Having the new defaulted method doesn't change things. In fact I can mark it as `@FunctionalInterface` if you want. Users can still use lambdas.

Yes, removing the default would mean it's not longer functional and hence users could not use lambdas. However, my latest thinking is that we should not remove the default implementation.  The default implementation's return value means the type doesn't conceptually have a base type its comparing to. 

Yes, we could achieve similar functionality by splitting into the two interfaces you mention. But, adding a defaulted method is cleaner: it doesn't require casts in the code and is more intuitive / more descriptive / more discoverable, IMHO.

Reply:
> Now that I realize it, this comment would basically boil down to: add a new method to `VarargMatcher` that defines its type.

This was my original approach. 

Currently, is a VarargMatcher is last argument, it is used to match against each additional element in a vararg parameter. Unfortunately, this leaves edgecases that's can't be handled.

We could repurpose `VarargMatcher` to be the second interface we need, fixing the implementation. But I'm still in favour of the defaulted method on `ArgumentMatcher`.

Reply:
See #2805 for an approach adding `type()` to `VarargMatcher`.

Though, if we go this way, we should remove the default impl.

Reply:
Okay, given the amount of issues this approach fixes and the fact that this can still be written as a lambda, I am okay with this approach. Thank you for writing down the considerations and trade-offs, that helped a lot in figuring out on how to proceed.